avoid calling no-op convert most of the time

vtjnash opened the pull request:
The basics of this change is altering this lowering, then doing the same in a few strategic places manually:
```julia
struct FC{FieldTypeA, S}
           a::FieldTypeA
           b
           c::Any
           d::Int
       end

julia> code_lowered(FC{Int,Int})
# before
1 ─ %1 = Core.fieldtype(#ctor-self#, 1)
│   %2 = Base.convert(%1, a)
│   %3 = Core.fieldtype(#ctor-self#, 3)
│   %4 = Base.convert(%3, c)
│   %5 = Core.fieldtype(#ctor-self#, 4)
│   %6 = Base.convert(%5, d)
│   %7 = %new(#ctor-self#, %2, b, %4, %6)
└──      return %7

# after
1 ── %1  = Core.fieldtype(#ctor-self#, 1)
│    %2  = a isa %1
└───       goto #3 if not %2
2 ──       @_6 = a
└───       goto #4
3 ──       @_6 = Base.convert(%1, a)
4 ┄─ %7  = @_6
│    %8  = Core.fieldtype(#ctor-self#, 3)
│    %9  = c isa %8
└───       goto #6 if not %9
5 ──       @_7 = c
└───       goto #7
6 ──       @_7 = Base.convert(%8, c)
7 ┄─ %14 = @_7
│    %15 = Core.fieldtype(#ctor-self#, 4)
│    %16 = d isa %15
└───       goto #9 if not %16
8 ──       @_8 = d
└───       goto #10
9 ──       @_8 = Base.convert(%15, d)
10 ┄ %21 = @_8
│    %22 = %new(#ctor-self#, %7, b, %14, %21)
└───       return %22

```
This is not quite as elegant, but dispatch has a notable cost for latency, so doing something which doesn't generate edges can improve the compiler (since it is then impossible for the user to introduce dispatch changes into it). This is enough of an apparent latency / memory improvement due to the simplification of the abstract model that I think it is worth doing.

In particular, this PR saves 2MB from the system image (over 2% of the data)

vtjnash committed:
syntax: minor cleanup of convert-for-type-decl code [NFC]

vtjnash committed:
remove redundant type-assert for global assignment

Makes a better error message too.
Saves about 0.1% (0.15MB) on the standard system image.

vtjnash committed:
avoid calling no-op convert most of the time

While these are free at runtime, they incur a small compile time cost to
create an enumerate the backedge to `convert(::Type{T}, ::T) where T`
which is shadowed by pretty much every `convert` method ever added
later, requiring careful re-checking for invalidations. This drops the
number of specializations of that method from over 1000 to under 100.

With the other commits, seems to save over 1 MB (1%) from the standard
system image.

vtjnash committed:
teach inference about splitting isa on the type

Sometime `fieldtype` returns Union{Types}, which we can split during the
if, to avoid union splitting on the convert call later when.

aviatesk commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -1710,22 +1710,34 @@ function abstract_call_builtin(interp::AbstractInterpreter, f::Builtin, (; fargs
                     end
                 end
             end
+            # try splitting type argument, based on value
+            if isdispatchelem(widenconst(a2)) && a3 isa Union && !has_free_typevars(a3) && !isa(rt, Const)
+                b = ssa_def_slot(fargs[3], sv)
+                if isa(b, SlotNumber)
+                    # !(x isa T) implies !(Type{a2} <: T)
+                    # TODO: complete splitting, based on which portions of the Union a3 for which isa_tfunc returns Const(true) or Const(false) instead of Bool
+                    elsetype = typesubtract(a3, Type{widenconst(a2)}, InferenceParams(interp).max_union_splitting)
+                    return Conditional(b, a3, elsetype)
+                end
+            end
```
It would be better to have a highlevel test case for this conditional type propagation, e.g.:
```julia
struct Issue49123
    x::Union{Int,Nothing}
end

@test Base.return_types((Int, Issue49123,)) do a2, x
    T = fieldtype(x, :x)
    if a2 isa T
        throw(a2)
    else
        return T
    end
end |>  only == Type{Nothing}
```

vtjnash committed:
avoid calling no-op convert in lowering when already isa

This necessitated removing the unreachable DCE pass, since it would skip
emitting SSA assignments, and then attempt to emit a label and then use
the value, which was not legal. That necessitated making sure the IR was
valid even when it was dead (a couple places tried to emit literal #f
into the IR, instead of (null) or skipping it).

vtjnash commented:
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

vtjnash commented:
@nanosoldier `runtests()`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/49123#issuecomment-1501022378) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/e819310_vs_1cf5091/report.md).

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/49123#issuecomment-1501022393) has completed - possible new issues were detected.
A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/ffca15a_vs_1cf5091/report.html).

ToucheSir commented:
Is there a way to detect whether one is in one of these synthesized constructors? The change is great, but I'm pretty sure it'll cause rampant type instability in Zygote because it introduces control flow in a common path (default constructors). As long as there's an easy way to detect default constructor calls using a `Method`/`MethodInstance`/lowered `CodeInfo`, this should not be hard to address downstream.
